cardano-dex-contracts-onchain/ErgoDex/PContracts/PVestingWithPeriod.hs

```diff
@@ -0,0 +1,130 @@
+
+instance PUnsafeLiftDecl VestingWithPeriodConfig where type PLifted VestingWithPeriodConfig = VWP.VestingWithPeriodConfig
+deriving via (DerivePConstantViaData VWP.VestingWithPeriodConfig VestingWithPeriodConfig) instance (PConstantDecl VWP.VestingWithPeriodConfig)
+
+instance PTryFrom PData (PAsData VestingWithPeriodConfig)
+
+vestingWithPeriodValidatorT :: ClosedTerm (VestingWithPeriodConfig :--> VestingWithPeriodRedeemer :--> PScriptContext :--> PBool)
+vestingWithPeriodValidatorT = plam $ \conf' redeemer' ctx' -> unTermCont $ do
+  let ctx = pfield @"txInfo" # ctx'
+
+  conf     <- pletFieldsC @'["vestingStart", "vestingPeriodDuration", "totalVested", "periodVested", "pkhs", "vestingAC"] conf'
+  redeemer <- pletFieldsC @'["vestingPeriodIdx"] redeemer'
+  txInfo   <- pletFieldsC @'["validRange", "signatories"] ctx
+  let
+      vestingStart          = pfromData $ getField @"vestingStart" conf
+      vestingPeriodDuration = pfromData $ getField @"vestingPeriodDuration" conf
+  
+      totalVested  = pfromData $ getField @"totalVested" conf
+      periodVested = pfromData $ getField @"periodVested" conf
+      pkhs         = pfromData $ getField @"pkhs" conf
+      vestingAC    = pfromData $ getField @"vestingAC" conf
+      
+      vestingPeriodIdx = pfromData $ getField @"vestingPeriodIdx" redeemer
+  
+      sigs = pfromData $ getField @"signatories" txInfo
+  
+  validRange <- tletUnwrap $ getField @"validRange" txInfo
+  let
+      periodAdditionalTime = pmultiply # vestingPeriodIdx # vestingPeriodDuration
+  
+      periodStartTime = periodAdditionalTime + vestingStart
+      validTime       = pbefore # periodStartTime # validRange
+      validSignature  = pall # (containsSignature' # sigs) # pkhs
+
+      maxPeriodsQty = pdiv # totalVested # periodVested
+      isLastPeriod  = maxPeriodsQty #<= vestingPeriodIdx
+  
+  correctReward <- 
+     tlet $ 
+         isLastPeriod #|| (checkRewardAndDatumCorrectness # ctx' # conf' # totalVested # periodVested # vestingPeriodIdx # vestingAC)
+  pure $ validTime
+       #&& validSignature 
+       #&& correctReward
+
+checkRewardAndDatumCorrectness :: Term s (PScriptContext :--> VestingWithPeriodConfig :--> PInteger :--> PInteger :--> PInteger :--> PAssetClass :--> PBool)
+checkRewardAndDatumCorrectness = 
+    plam $ \ctx prevCfg totalVested periodVested periodId vestingAC -> unTermCont $ do
+        let
+            selfOutputsList = getContinuingOutputs # ctx
+            selfOutput      = phead # selfOutputsList
+            selfValue       = pfield @"value" # selfOutput
+        txOutDatum <- tletField @"datum" selfOutput
+
+        POutputDatum txOutOutputDatum <- pmatchC txOutDatum
+
+        rawDatum <- tletField @"outputDatum" txOutOutputDatum
+
+        PDatum vestingDatumRaw <- pmatchC rawDatum
+        newVestingConfig <- tletUnwrap $ ptryFromData @(VestingWithPeriodConfig) $ vestingDatumRaw
+        let    
+            correctOutQty   = totalVested - (periodId * periodVested) 
```

Comment:
So basically it's possible to withdraw as many tokens as is available in the UTxO in total?

Reply:
after ending period with id N, you can withdraw N * periodVested tokens